pkg/ring/basic_lifecycler.go

```diff
@@ -0,0 +1,474 @@
+package ring
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cortexproject/cortex/pkg/ring/kv"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+type BasicLifecyclerDelegate interface {
+	// OnRingInstanceRegister is called while the lifecycler is registering the
+	// instance within the ring and should return the state and set of tokens to
+	// use for the instance itself.
+	OnRingInstanceRegister(lifecycler *BasicLifecycler, ringDesc Desc, instanceExists bool, instanceID string, instanceDesc IngesterDesc) (IngesterState, Tokens)
+
+	// OnRingInstanceTokens is called once the instance tokens are set and are
+	// stable within the ring (honoring the observe period, if set).
+	OnRingInstanceTokens(lifecycler *BasicLifecycler, tokens Tokens)
+
+	// OnRingInstanceStopping is called while the lifecycler is stopping. The lifecycler
+	// will continue to hearbeat the ring the this function is executing and will proceed
+	// to unregister the instance from the ring only after this function has returned.
+	OnRingInstanceStopping(lifecycler *BasicLifecycler)
+}
+
+type OnRingInstanceRegister func(lifecycler *BasicLifecycler, ringDesc Desc, instanceExists bool, instanceID string, instanceDesc IngesterDesc) (IngesterState, Tokens)
+type OnRingInstanceTokensChanged func(lifecycler *BasicLifecycler, tokens Tokens)
+type OnRingInstanceStopping func(lifecycler *BasicLifecycler)
+
+type BasicLifecyclerConfig struct {
+	InstanceID          string
+	InstanceAddr        string
+	InstancePort        int
```

Comment:
Would it make sense to put addr and port together into a field? Lifecycler doesn't care about port specifically, and only writes this to ring.

Also `Instance` prefix seems excessive.

Reply:
Agree that addr and port can be merged together. I've also removed the `Instance` prefix even if I usually prefer clarity vs compacted options.